src/rust/src/x509/sign.rs

```diff
@@ -119,14 +119,88 @@ fn identify_hash_type(
     }
 }
 
+fn compute_pss_salt_length<'p>(
+    py: pyo3::Python<'p>,
+    private_key: &'p pyo3::PyAny,
+    hash_algorithm: &'p pyo3::PyAny,
+    rsa_padding: &'p pyo3::PyAny,
+) -> pyo3::PyResult<u16> {
+    let padding_mod = py.import(pyo3::intern!(
+        py,
+        "cryptography.hazmat.primitives.asymmetric.padding"
+    ))?;
+    let maxlen = padding_mod.getattr(pyo3::intern!(py, "_MaxLength"))?;
+    let digestlen = padding_mod.getattr(pyo3::intern!(py, "_DigestLength"))?;
+    let py_saltlen = rsa_padding.getattr(pyo3::intern!(py, "_salt_length"))?;
+    if py_saltlen.is_instance(maxlen)? {
+        padding_mod
+            .getattr(pyo3::intern!(py, "calculate_max_pss_salt_length"))?
+            .call1((private_key, hash_algorithm))?
+            .extract::<u16>()
+    } else if py_saltlen.is_instance(digestlen)? {
+        hash_algorithm
+            .getattr(pyo3::intern!(py, "digest_size"))?
+            .extract::<u16>()
+    } else if py_saltlen.is_instance(py.get_type::<pyo3::types::PyLong>())? {
+        py_saltlen.extract::<u16>()
+    } else {
+        Err(pyo3::exceptions::PyTypeError::new_err(
+            "salt_length must be an int, MaxLength, or DigestLength.",
+        ))
```

Comment:
I feel like this validation should be in `PSS.__init__`, not here. I say just make it an `else` and do the `extract`. If you want this validation, seperate PR to init IMO.

Reply:
`PSS.__init__` does enforce: https://github.com/pyca/cryptography/blob/3b8cb2b7337a7dc3a004555fd4bfa032e35a34af/src/cryptography/hazmat/primitives/asymmetric/padding.py#L47-L53
but it allows `AUTO`, which has no meaning here (it's only valid in verification contexts for recovering salt length without specifying it).

So we can do the else extract but then if they've passed auto this will fail.

Reply:
😭 ok, I forgot about AUTO